File: cli/src/core/MetadataManager.test.ts
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import type { ManifestEntry } from "./KBTypes.js";
import { MetadataManager } from "./MetadataManager.js";

function makeTmpDir(): string {
	const dir = join(tmpdir(), `kb-test-${Date.now()}-${Math.random().toString(36).slice(2)}`);
	mkdirSync(dir, { recursive: true });
	return dir;
}

function makeEntry(overrides: Partial<ManifestEntry> = {}): ManifestEntry {
	return {
		path: "main/test-abc12345.md",
		fileId: "abc12345",
		type: "commit",
		fingerprint: "sha256:deadbeef",
		source: { commitHash: "abc12345", branch: "main" },
		...overrides,
	};
}

function rmrf(dir: string): void {
	const { rmSync } = require("node:fs");
	try {
		rmSync(dir, { recursive: true, force: true });
	} catch {
		/* ignore */
	}
}

describe("MetadataManager", () => {
	let jolliDir: string;
	let manager: MetadataManager;

	beforeEach(() => {
		jolliDir = join(makeTmpDir(), ".jolli");
		manager = new MetadataManager(jolliDir);
	});

	afterEach(() => {
		rmrf(join(jolliDir, ".."));
	});

	describe("ensure", () => {
		it("creates jolli dir and default files", () => {
			manager.ensure();
			expect(existsSync(jolliDir)).toBe(true);
			expect(existsSync(join(jolliDir, "manifest.json"))).toBe(true);
			expect(existsSync(join(jolliDir, "branches.json"))).toBe(true);
			expect(existsSync(join(jolliDir, "config.json"))).toBe(true);
		});

		it("is idempotent", () => {
			manager.ensure();
			manager.ensure();
			expect(existsSync(jolliDir)).toBe(true);
		});

		it("does not overwrite existing files", () => {
			manager.ensure();
			manager.updateManifest(makeEntry());
			manager.ensure();
			expect(manager.readManifest().files).toHaveLength(1);
		});
	});

	describe("manifest", () => {
		beforeEach(() => manager.ensure());

		it("reads empty manifest when fresh", () => {
			const manifest = manager.readManifest();
			expect(manifest.version).toBe(1);
			expect(manifest.files).toHaveLength(0);
		});

		it("adds new entry", () => {
			manager.updateManifest(makeEntry());
			expect(manager.readManifest().files).toHaveLength(1);
			expect(manager.readManifest().files[0].fileId).toBe("abc12345");
		});

		it("replaces entry with same fileId", () => {
			manager.updateManifest(makeEntry({ path: "main/old.md" }));
			manager.updateManifest(makeEntry({ path: "main/new.md" }));
			const files = manager.readManifest().files;
			expect(files).toHaveLength(1);
			expect(files[0].path).toBe("main/new.md");
		});

		it("keeps entries with different fileId", () => {
			manager.updateManifest(makeEntry({ fileId: "aaa" }));
			manager.updateManifest(makeEntry({ fileId: "bbb" }));
			expect(manager.readManifest().files).toHaveLength(2);
		});

		it("removes entry by fileId", () => {
			manager.updateManifest(makeEntry({ fileId: "aaa" }));
			manager.updateManifest(makeEntry({ fileId: "bbb" }));
			expect(manager.removeFromManifest("aaa")).toBe(true);
			expect(manager.readManifest().files).toHaveLength(1);
		});

		it("returns false when removing nonexistent fileId", () => {
			expect(manager.removeFromManifest("nonexistent")).toBe(false);
		});

		it("findByPath returns matching entry", () => {
			manager.updateManifest(makeEntry({ path: "main/test.md", fileId: "abc" }));
			expect(manager.findByPath("main/test.md")?.fileId).toBe("abc");
		});

		it("findById returns matching entry", () => {
			manager.updateManifest(makeEntry({ fileId: "xyz" }));
			expect(manager.findById("xyz")).toBeDefined();
		});

		it("updatePath changes path", () => {
			manager.updateManifest(makeEntry({ path: "old/path.md", fileId: "abc" }));
			expect(manager.updatePath("abc", "new/path.md")).toBe(true);
			expect(manager.findById("abc")?.path).toBe("new/path.md");
		});
	});

	describe("branch mapping", () => {
		beforeEach(() => manager.ensure());

		it("creates new mapping", () => {
			const folder = manager.resolveFolderForBranch("feature/login");
			expect(folder).toBe("feature-login");
			expect(manager.listBranchMappings()).toHaveLength(1);
		});

		it("returns existing mapping on second call", () => {
			manager.resolveFolderForBranch("feature/foo");
			manager.resolveFolderForBranch("feature/foo");
			expect(manager.listBranchMappings()).toHaveLength(1);
		});

		it("renames branch folder and updates manifest", () => {
			manager.resolveFolderForBranch("feature/old");
			manager.updateManifest(makeEntry({ path: "feature-old/a.md", fileId: "a" }));
			manager.updateManifest(makeEntry({ path: "feature-old/b.md", fileId: "b" }));

			const count = manager.renameBranchFolder("feature-old", "feature-new");
			expect(count).toBe(2);
			expect(manager.findById("a")?.path).toBe("feature-new/a.md");
			expect(manager.findById("b")?.path).toBe("feature-new/b.md");
		});

		it("removes branch folder from manifest and branches", () => {
			manager.resolveFolderForBranch("feature/delete-me");
			manager.updateManifest(makeEntry({ path: "feature-delete-me/a.md", fileId: "a" }));
			manager.updateManifest(makeEntry({ path: "main/b.md", fileId: "b" }));

			const removed = manager.removeBranchFolder("feature-delete-me");
			expect(removed).toBe(1);
			expect(manager.readManifest().files).toHaveLength(1);
			expect(manager.findById("b")).toBeDefined();
		});
	});

	describe("transcodeBranchName", () => {
		it("replaces forward slash", () => {
			expect(MetadataManager.transcodeBranchName("feature/login")).toBe("feature-login");
		});

		it("replaces backslash", () => {
			expect(MetadataManager.transcodeBranchName("user\\foo")).toBe("user-foo");
		});

		it("replaces double dot with double dash", () => {
			expect(MetadataManager.transcodeBranchName("refs..heads")).toBe("refs--heads");
		});

		it("collapses consecutive dashes", () => {
			expect(MetadataManager.transcodeBranchName("a///b")).toBe("a-b");
		});

		it("trims leading/trailing dots and dashes", () => {
			expect(MetadataManager.transcodeBranchName(".leading")).toBe("leading");
			expect(MetadataManager.transcodeBranchName("trailing-")).toBe("trailing");
		});

		it("returns default for empty result", () => {
			expect(MetadataManager.transcodeBranchName("/")).toBe("default");
		});

		it("simple names pass through", () => {
			expect(MetadataManager.transcodeBranchName("main")).toBe("main");
		});
	});

	describe("config", () => {
		beforeEach(() => manager.ensure());

		it("reads defaults", () => {
			const config = manager.readConfig();
			expect(config.version).toBe(1);
			expect(config.sortOrder).toBe("date");
		});

		it("saves and reads back", () => {
			manager.saveConfig({ version: 1, sortOrder: "name" });
			expect(manager.readConfig().sortOrder).toBe("name");
		});
	});

	describe("migration state", () => {
		beforeEach(() => manager.ensure());

		it("returns null when no migration", () => {
			expect(manager.readMigrationState()).toBeNull();
		});

		it("saves and reads back", () => {
			manager.saveMigrationState({ status: "completed", totalEntries: 5, migratedEntries: 5 });
			const state = manager.readMigrationState();
			expect(state?.status).toBe("completed");
			expect(state?.migratedEntries).toBe(5);
		});
	});

	describe("reconcile", () => {
		beforeEach(() => manager.ensure());

		it("keeps manifest entry when file is deleted (avoids data loss from transient errors)", () => {
			const kbRoot = join(jolliDir, "..");
			mkdirSync(join(kbRoot, "main"), { recursive: true });
			writeFileSync(join(kbRoot, "main/test.md"), "# Test");
			manager.updateManifest(makeEntry({ path: "main/test.md", fileId: "abc", fingerprint: "fp1" }));

			// Delete the file
			require("node:fs").unlinkSync(join(kbRoot, "main/test.md"));

			const fixed = manager.reconcile(kbRoot);
			expect(fixed).toBe(0);
			// Entry is preserved to avoid data loss from transient filesystem errors
			expect(manager.findById("abc")).toBeDefined();
		});

		it("updates path when file is moved (fingerprint match)", () => {
			const kbRoot = join(jolliDir, "..");
			mkdirSync(join(kbRoot, "other"), { recursive: true });
			const content = "# Moved file content";
			const fp = MetadataManager.sha256(content);

			manager.updateManifest(makeEntry({ path: "main/test.md", fileId: "abc", fingerprint: fp }));
			writeFileSync(join(kbRoot, "other/test.md"), content);

			const fixed = manager.reconcile(kbRoot);
			expect(fixed).toBe(1);
			expect(manager.findById("abc")?.path).toBe("other/test.md");
		});

		it("no changes when files are in place", () => {
			const kbRoot = join(jolliDir, "..");
			mkdirSync(join(kbRoot, "main"), { recursive: true });
			writeFileSync(join(kbRoot, "main/test.md"), "# Test");
			manager.updateManifest(makeEntry({ path: "main/test.md", fileId: "abc" }));

			expect(manager.reconcile(kbRoot)).toBe(0);
		});
	});

	describe("resilience", () => {
		it("reads empty manifest on corrupted file", () => {
			mkdirSync(jolliDir, { recursive: true });
			writeFileSync(join(jolliDir, "manifest.json"), "not json!!!");
			expect(manager.readManifest().files).toHaveLength(0);
		});

		it("reads defaults on corrupted config", () => {
			mkdirSync(jolliDir, { recursive: true });
			writeFileSync(join(jolliDir, "config.json"), "{bad json");
			expect(manager.readConfig().sortOrder).toBe("date");
		});
	});
});

File: intellij/src/test/kotlin/ai/jolli/jollimemory/core/MetadataManagerTest.kt
package ai.jolli.jollimemory.core

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path

class MetadataManagerTest {

    @TempDir
    lateinit var tempDir: Path

    private lateinit var jolliDir: Path
    private lateinit var manager: MetadataManager
    private val gson: Gson = GsonBuilder().setPrettyPrinting().create()

    @BeforeEach
    fun setUp() {
        jolliDir = tempDir.resolve(".jolli")
        manager = MetadataManager(jolliDir)
    }

    private fun makeEntry(
        path: String = "main/abc12345-test.md",
        fileId: String = "abc12345",
        type: String = "commit",
        fingerprint: String = "sha256:deadbeef",
    ) = ManifestEntry(
        path = path,
        fileId = fileId,
        type = type,
        fingerprint = fingerprint,
        source = ManifestSource(commitHash = "abc12345", branch = "main"),
    )

    // ── ensure ─────────────────────────────────────────────────────────────

    @Nested
    inner class Ensure {
        @Test
        fun `creates jolli dir and default files`() {
            manager.ensure()

            Files.isDirectory(jolliDir) shouldBe true
            Files.exists(jolliDir.resolve("manifest.json")) shouldBe true
            Files.exists(jolliDir.resolve("branches.json")) shouldBe true
            Files.exists(jolliDir.resolve("config.json")) shouldBe true
        }

        @Test
        fun `is idempotent — calling twice does not fail`() {
            manager.ensure()
            manager.ensure()

            Files.isDirectory(jolliDir) shouldBe true
        }

        @Test
        fun `does not overwrite existing files`() {
            manager.ensure()
            // Modify manifest
            manager.updateManifest(makeEntry())

            // Call ensure again — should not reset manifest
            manager.ensure()
            manager.readManifest().files shouldHaveSize 1
        }
    }

    // ── Manifest ───────────────────────────────────────────────────────────

    @Nested
    inner class ManifestOps {
        @BeforeEach
        fun init() { manager.ensure() }

        @Test
        fun `readManifest returns empty manifest when fresh`() {
            val manifest = manager.readManifest()
            manifest.version shouldBe 1
            manifest.files.shouldBeEmpty()
        }

        @Test
        fun `updateManifest adds new entry`() {
            manager.updateManifest(makeEntry())
            manager.readManifest().files shouldHaveSize 1
            manager.readManifest().files[0].fileId shouldBe "abc12345"
        }

        @Test
        fun `updateManifest replaces entry with same fileId`() {
            manager.updateManifest(makeEntry(path = "main/old.md"))
            manager.updateManifest(makeEntry(path = "main/new.md"))

            val files = manager.readManifest().files
            files shouldHaveSize 1
            files[0].path shouldBe "main/new.md"
        }

        @Test
        fun `updateManifest keeps entries with different fileId`() {
            manager.updateManifest(makeEntry(fileId = "aaa"))
            manager.updateManifest(makeEntry(fileId = "bbb"))

            manager.readManifest().files shouldHaveSize 2
        }

        @Test
        fun `removeFromManifest removes existing entry`() {
            manager.updateManifest(makeEntry(fileId = "aaa"))
            manager.updateManifest(makeEntry(fileId = "bbb"))

            manager.removeFromManifest("aaa") shouldBe true
            manager.readManifest().files shouldHaveSize 1
            manager.readManifest().files[0].fileId shouldBe "bbb"
        }

        @Test
        fun `removeFromManifest returns false when fileId not found`() {
            manager.removeFromManifest("nonexistent") shouldBe false
        }

        @Test
        fun `findByPath returns matching entry`() {
            manager.updateManifest(makeEntry(path = "main/test.md", fileId = "abc"))
            manager.findByPath("main/test.md")?.fileId shouldBe "abc"
        }

        @Test
        fun `findByPath returns null when not found`() {
            manager.findByPath("nonexistent") shouldBe null
        }

        @Test
        fun `findById returns matching entry`() {
            manager.updateManifest(makeEntry(fileId = "xyz"))
            manager.findById("xyz") shouldNotBe null
        }

        @Test
        fun `findById returns null when not found`() {
            manager.findById("nonexistent") shouldBe null
        }

        @Test
        fun `updatePath changes path for existing entry`() {
            manager.updateManifest(makeEntry(path = "old/path.md", fileId = "abc"))

            manager.updatePath("abc", "new/path.md") shouldBe true
            manager.findById("abc")?.path shouldBe "new/path.md"
        }

        @Test
        fun `updatePath returns false for nonexistent fileId`() {
            manager.updatePath("nonexistent", "new.md") shouldBe false
        }
    }

    // ── Branch mapping ─────────────────────────────────────────────────────

    @Nested
    inner class BranchMappingOps {
        @BeforeEach
        fun init() { manager.ensure() }

        @Test
        fun `resolveFolderForBranch creates new mapping`() {
            val folder = manager.resolveFolderForBranch("feature/jolli-400")
            folder shouldBe "feature-jolli-400"

            val mappings = manager.listBranchMappings()
            mappings shouldHaveSize 1
            mappings[0].branch shouldBe "feature/jolli-400"
            mappings[0].folder shouldBe "feature-jolli-400"
        }

        @Test
        fun `resolveFolderForBranch returns existing mapping on second call`() {
            manager.resolveFolderForBranch("feature/foo")
            val folder = manager.resolveFolderForBranch("feature/foo")

            folder shouldBe "feature-foo"
            manager.listBranchMappings() shouldHaveSize 1
        }

        @Test
        fun `resolveFolderForBranch handles multiple branches`() {
            manager.resolveFolderForBranch("main")
            manager.resolveFolderForBranch("feature/foo")
            manager.resolveFolderForBranch("bugfix/bar")

            manager.listBranchMappings() shouldHaveSize 3
        }

        @Test
        fun `updateBranchMapping changes branch for existing folder`() {
            manager.resolveFolderForBranch("feature/old")
            manager.updateBranchMapping("feature-old", "feature/new")

            val mappings = manager.listBranchMappings()
            mappings[0].branch shouldBe "feature/new"
            mappings[0].folder shouldBe "feature-old"
        }
    }

    // ── Branch name transcoding ────────────────────────────────────────────

    @Nested
    inner class TranscodeBranchName {
        @Test
        fun `replaces forward slash`() {
            MetadataManager.transcodeBranchName("feature/jolli-400") shouldBe "feature-jolli-400"
        }

        @Test
        fun `replaces backslash`() {
            MetadataManager.transcodeBranchName("user\\foo") shouldBe "user-foo"
        }

        @Test
        fun `replaces colon`() {
            MetadataManager.transcodeBranchName("refs:heads") shouldBe "refs-heads"
        }

        @Test
        fun `replaces asterisk`() {
            MetadataManager.transcodeBranchName("feature*test") shouldBe "feature-test"
        }

        @Test
        fun `replaces question mark`() {
            MetadataManager.transcodeBranchName("what?this") shouldBe "what-this"
        }

        @Test
        fun `replaces tilde`() {
            MetadataManager.transcodeBranchName("branch~1") shouldBe "branch-1"
        }

        @Test
        fun `replaces caret`() {
            MetadataManager.transcodeBranchName("branch^2") shouldBe "branch-2"
        }

        @Test
        fun `replaces double dot with double dash`() {
            MetadataManager.transcodeBranchName("refs..heads") shouldBe "refs--heads"
        }

        @Test
        fun `collapses consecutive dashes`() {
            MetadataManager.transcodeBranchName("a///b") shouldBe "a-b"
        }

        @Test
        fun `trims leading dot and dash`() {
            MetadataManager.transcodeBranchName(".leading") shouldBe "leading"
            MetadataManager.transcodeBranchName("-leading") shouldBe "leading"
        }

        @Test
        fun `trims trailing dot and dash`() {
            MetadataManager.transcodeBranchName("trailing.") shouldBe "trailing"
            MetadataManager.transcodeBranchName("trailing-") shouldBe "trailing"
        }

        @Test
        fun `simple branch names pass through`() {
            MetadataManager.transcodeBranchName("main") shouldBe "main"
            MetadataManager.transcodeBranchName("develop") shouldBe "develop"
        }

        @Test
        fun `returns default for empty result`() {
            MetadataManager.transcodeBranchName("/") shouldBe "default"
            MetadataManager.transcodeBranchName("...") shouldBe "default"
        }

        @Test
        fun `complex branch name`() {
            MetadataManager.transcodeBranchName("user/name/feature/JIRA-123") shouldBe "user-name-feature-JIRA-123"
        }
    }

    // ── Index ──────────────────────────────────────────────────────────────

    @Nested
    inner class IndexOps {
        @BeforeEach
        fun init() { manager.ensure() }

        @Test
        fun `readIndex returns null when no index exists`() {
            manager.readIndex() shouldBe null
        }

        @Test
        fun `writeIndex and readIndex round-trip`() {
            val index = SummaryIndex(
                version = 3,
                entries = listOf(
                    SummaryIndexEntry(
                        commitHash = "abc123",
                        commitMessage = "test",
                        commitDate = "2026-01-01T00:00:00Z",
                        branch = "main",
                        generatedAt = "2026-01-01T00:00:00Z",
                    )
                ),
            )
            manager.writeIndex(index)

            val read = manager.readIndex()
            read shouldNotBe null
            read!!.entries shouldHaveSize 1
            read.entries[0].commitHash shouldBe "abc123"
        }

        @Test
        fun `rebuildIndex scans summary JSON files`() {
            val summariesDir = jolliDir.resolve("summaries")
            Files.createDirectories(summariesDir)

            val summary = CommitSummary(
                commitHash = "abc123",
                commitMessage = "test commit",
                commitAuthor = "Alice",
                commitDate = "2026-01-15T10:00:00Z",
                branch = "main",
                generatedAt = "2026-01-15T10:00:00Z",
                topics = listOf(
                    TopicSummary("t1", "trigger", "response", "decisions")
                ),
                stats = DiffStats(3, 10, 5),
            )
            Files.writeString(summariesDir.resolve("abc123.json"), gson.toJson(summary))

            // Add to manifest so rebuildIndex finds it
            manager.updateManifest(makeEntry(fileId = "abc123", type = "commit"))

            val index = manager.rebuildIndex(summariesDir)
            index.entries shouldHaveSize 1
            index.entries[0].commitHash shouldBe "abc123"
            index.entries[0].commitMessage shouldBe "test commit"
            index.entries[0].topicCount shouldBe 1
            index.entries[0].diffStats?.filesChanged shouldBe 3

            // Should also persist
            manager.readIndex()!!.entries shouldHaveSize 1
        }

        @Test
        fun `rebuildIndex skips missing summary files`() {
            manager.updateManifest(makeEntry(fileId = "missing", type = "commit"))

            val summariesDir = jolliDir.resolve("summaries")
            Files.createDirectories(summariesDir)

            val index = manager.rebuildIndex(summariesDir)
            index.entries.shouldBeEmpty()
        }

        @Test
        fun `rebuildIndex ignores non-commit manifest entries`() {
            manager.updateManifest(makeEntry(fileId = "plan1", type = "plan"))

            val summariesDir = jolliDir.resolve("summaries")
            Files.createDirectories(summariesDir)

            val index = manager.rebuildIndex(summariesDir)
            index.entries.shouldBeEmpty()
        }
    }

    // ── Config ─────────────────────────────────────────────────────────────

    @Nested
    inner class ConfigOps {
        @BeforeEach
        fun init() { manager.ensure() }

        @Test
        fun `readConfig returns defaults when fresh`() {
            val config = manager.readConfig()
            config.version shouldBe 1
            config.sortOrder shouldBe "date"
        }

        @Test
        fun `saveConfig and readConfig round-trip`() {
            manager.saveConfig(KBConfig(sortOrder = "name"))

            val config = manager.readConfig()
            config.sortOrder shouldBe "name"
        }
    }

    // ── Resilience ─────────────────────────────────────────────────────────

    @Nested
    inner class Resilience {
        @Test
        fun `readManifest returns empty manifest on corrupted file`() {
            Files.createDirectories(jolliDir)
            Files.writeString(jolliDir.resolve("manifest.json"), "not json!!!")

            manager.readManifest().files.shouldBeEmpty()
        }

        @Test
        fun `readConfig returns defaults on corrupted file`() {
            Files.createDirectories(jolliDir)
            Files.writeString(jolliDir.resolve("config.json"), "{bad json")

            manager.readConfig().sortOrder shouldBe "date"
        }
    }

    // ── Branch folder rename ──────────────────────────────────────────────

    @Nested
    inner class RenameBranchFolder {
        @BeforeEach
        fun init() { manager.ensure() }

        @Test
        fun `renames branch folder in branches json and manifest`() {
            manager.resolveFolderForBranch("feature/old")
            manager.updateManifest(makeEntry(path = "feature-old/abc.md", fileId = "abc"))
            manager.updateManifest(makeEntry(path = "feature-old/def.md", fileId = "def"))

            val count = manager.renameBranchFolder("feature-old", "feature-new")
            count shouldBe 2

            // Manifest paths updated
            manager.findById("abc")?.path shouldBe "feature-new/abc.md"
            manager.findById("def")?.path shouldBe "feature-new/def.md"

            // Branches json updated
            val mappings = manager.listBranchMappings()
            mappings.any { it.folder == "feature-new" } shouldBe true
            mappings.any { it.folder == "feature-old" } shouldBe false
        }

        @Test
        fun `does not affect other branch entries`() {
            manager.updateManifest(makeEntry(path = "main/x.md", fileId = "x"))
            manager.updateManifest(makeEntry(path = "feature-old/y.md", fileId = "y"))

            manager.renameBranchFolder("feature-old", "feature-new")

            manager.findById("x")?.path shouldBe "main/x.md"
            manager.findById("y")?.path shouldBe "feature-new/y.md"
        }
    }

    // ── Remove branch folder ──────────────────────────────────────────────

    @Nested
    inner class RemoveBranchFolder {
        @BeforeEach
        fun init() { manager.ensure() }

        @Test
        fun `removes branch folder entries from manifest and branches json`() {
            manager.resolveFolderForBranch("feature/delete-me")
            manager.updateManifest(makeEntry(path = "feature-delete-me/a.md", fileId = "a"))
            manager.updateManifest(makeEntry(path = "feature-delete-me/b.md", fileId = "b"))
            manager.updateManifest(makeEntry(path = "main/c.md", fileId = "c"))

            val removed = manager.removeBranchFolder("feature-delete-me")
            removed shouldBe 2

            manager.readManifest().files shouldHaveSize 1
            manager.findById("c") shouldNotBe null
            manager.listBranchMappings().any { it.folder == "feature-delete-me" } shouldBe false
        }
    }

    // ── Reconciliation ────────────────────────────────────────────────────

    @Nested
    inner class Reconciliation {
        @BeforeEach
        fun init() { manager.ensure() }

        @Test
        fun `removes manifest entry when file is deleted`() {
            val kbRoot = tempDir.resolve("kb")
            Files.createDirectories(kbRoot.resolve("main"))
            val file = kbRoot.resolve("main/test.md")
            Files.writeString(file, "# Test")

            manager.updateManifest(makeEntry(path = "main/test.md", fileId = "abc", fingerprint = "fp1"))

            // Delete the file
            Files.delete(file)

            val fixed = manager.reconcile(kbRoot)
            fixed shouldBe 1
            manager.findById("abc") shouldBe null
        }

        @Test
        fun `updates path when file is moved (same fingerprint)`() {
            val kbRoot = tempDir.resolve("kb")
            Files.createDirectories(kbRoot.resolve("main"))
            Files.createDirectories(kbRoot.resolve("other"))
            val content = "# Moved file content"
            val fp = ai.jolli.jollimemory.core.FolderStorage.sha256(content)

            // File originally at main/test.md
            manager.updateManifest(makeEntry(path = "main/test.md", fileId = "abc", fingerprint = fp))

            // File moved to other/test.md
            Files.writeString(kbRoot.resolve("other/test.md"), content)

            val fixed = manager.reconcile(kbRoot)
            fixed shouldBe 1
            manager.findById("abc")?.path shouldBe "other/test.md"
        }

        @Test
        fun `no changes when files are in place`() {
            val kbRoot = tempDir.resolve("kb")
            Files.createDirectories(kbRoot.resolve("main"))
            Files.writeString(kbRoot.resolve("main/test.md"), "# Test")

            manager.updateManifest(makeEntry(path = "main/test.md", fileId = "abc"))

            val fixed = manager.reconcile(kbRoot)
            fixed shouldBe 0
        }
    }
}
